beaker/application.py

```diff
@@ -1,418 +1,963 @@
+                self.deregister_bare_method(existing_action.action)
+            self.bare_actions[for_action] = OnCompleteAction(
+                action=sub, call_config=call_config
+            )
 
-        self.on_create: Optional[HandlerFunc] = (
-            all_creates.pop() if len(all_creates) == 1 else None
-        )
-        self.on_update: Optional[HandlerFunc] = (
-            all_updates.pop() if len(all_updates) == 1 else None
+    def deregister_bare_method(
+        self,
+        action_name_or_reference: OnCompleteActionName | SubroutineFnWrapper,
+        /,
+    ) -> None:
+        if isinstance(action_name_or_reference, str):
+            self.bare_actions.pop(cast(OnCompleteActionName, action_name_or_reference))
+        else:
+            method = action_name_or_reference
+            _remove_first_match(self.bare_actions, lambda _, v: v.action is method)
+
+    @overload
+    def external(
+        self,
+        fn: HandlerFunc,
+        /,
+    ) -> ABIReturnSubroutine:
+        ...
+
+    @overload
+    def external(
+        self,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorFuncType:
+        ...
+
+    def external(
+        self,
+        fn: HandlerFunc | None = None,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorResultType | DecoratorFuncType:
+        """
+        Add the method decorated to be handled as an ABI method for the Application
+
+        Args:
+            fn: The function being wrapped.
+            method_config:  <TODO>
+            name: Name of ABI method. If not set, name of the python method will be used.
+                Useful for method overriding.
+            authorize: a subroutine with input of ``Txn.sender()`` and output uint64
+                interpreted as allowed if the output>0.
+            bare:
+            read_only: Mark a method as callable with no fee using dryrun or simulate
+            override:
+
+        Returns:
+            The original method with additional elements set in it's
+            :code:`__handler_config__` attribute
+        """
+
+        def decorator(func: HandlerFunc) -> SubroutineFnWrapper | ABIReturnSubroutine:
+            sig = inspect.signature(func)
+            nonlocal bare
+            if bare is None:
+                bare = not sig.parameters
+
+            if bare and read_only:
+                raise ValueError("read_only has no effect on bare methods")
```

Comment:
I think this should be allowed for noops? others like update/delete/etc shouldnt since they'd cause an effect but noop method with no args should be ok

Reply:
So this check is in there because if the user specifies `read_only=True`, that only gets put into the app-spec if it's an ABI method, not if it's a bare method.